ruma-identifiers/src/mxc.rs

```diff
@@ -0,0 +1,119 @@
+//! Matrix-spec compliant mxc:// urls.
+
+use std::{
+    convert::TryFrom,
+    fmt::{self, Display},
+    str::FromStr,
+};
+
+use ruma_identifiers_validation::mxc::validate;
+
+use crate::{Error, ServerName, ServerNameBox};
+
+/// Matrix-spec compliant mxc:// urls.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Mxc {
+    server_name: ServerNameBox,
+    media_id: Box<str>,
+}
+
+impl Mxc {
+    /// Returns the media ID of this mxc://.
+    pub fn media_id(&self) -> &str {
+        &self.media_id
+    }
+
+    /// Returns the server name of this mxc://.
+    pub fn server_name(&self) -> &ServerName {
+        self.server_name.as_ref()
+    }
+}
+
+impl fmt::Debug for Mxc {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.to_string())
```

Comment:
I don't like that this allocates a temporary string when `Display` doesn't. Can you add an inherent `_fmt` that does what `Display::fmt` does now and is called by both `Debug` and `Display`?